scylla-gdb.py

```diff
@@ -5064,9 +5061,35 @@ def invoke(self, ptr):
         if m is None:
             raise ValueError("Failed to extract type name from symbol name `{}'".format(symbol_name))
 
-        actual_type = gdb.lookup_type(m.group(1)).pointer()
-        return ptr.reinterpret_cast(actual_type)
+        return m
+
+    def invoke(self, ptr):
+        if not isinstance(ptr, gdb.Value):
+            ptr = gdb.parse_and_eval(ptr)
+
+        m = self._check_vptr(ptr)
+
+        actual_type = gdb.lookup_type(m.group(1))
+        actual_type_ptr = actual_type.pointer()
+
+        if int(m.group(2)) == 16:
+            return ptr.reinterpret_cast(actual_type_ptr)
 
+        # We are most likely dealing with multiple inheritance and a pointer to a
+        # non-first base-class type.
+        base_class_field = actual_type.fields()[0]
+        assert(base_class_field.is_base_class)
+        base_classes = list(base_class_field.type.fields())
+
+        # The pointer is surely not to the first base-class, we would have found
+        # the expected m.group(2) == 16 offset otherwise.
+        for bc in base_classes[1:]:
+            speculative_ptr = gdb.Value(int(ptr) - int(bc.bitpos / 8))
+            m = self._check_vptr(speculative_ptr)
+            if int(m.group(2)) == 16:
+                return speculative_ptr.reinterpret_cast(actual_type_ptr)
+
```

Comment:
I have a feeling it could be coded in a more neutral way, but it isn't worth the effort to dig in for all the details (and the feeling may well be wrong).